### ISSUE TITLE
`eqx.tree_at` fails with `TypeError: Operation undefined, {x} is not a leaf of the pytree` when custom pytree attributes are modified within `__init__` 

### ISSUE THREAD

**Issue by nicolasJouvin** [patrick-kidger/equinox#711]
Hi, I'm posting this on Equinox but I don't in which extent it is a pure JAX-related issue. I have a MWE with the following custom class registered as a pytree where the `__init__` function pre-process one of the dynamic attribute (`dict`) of the class. JAX flattening and tree reconstruction seems to work as expected, but I get an error when trying to use `eqx.tree_at` to replace another leaf (`attr`) of the pytree. The error I get seems to be unrelated to the actual `attr` value I'm trying to modify, but rather related to the `__init__` and the fact that I "force" another tree leaf `self.dict["a"]` to some value (42). 

I understand this is clearly happening when reconstructing the new pytree in `eqx.tree_at` but I have trouble to understand if this an intended behavior of JAX or equinox ? I don't see what prevents me to do such things in a custom pytree initialization.

Thanks in advance for any answers or related issues that could help me understand this better.

```
import jax
import equinox as eqx


class CustomPytree:
    def __init__(self, dict, attr, static):
        self.dict = dict
        # preprocess attribute -> always set key "a" to 42
        self.dict["a"] = 42
        self.attr = attr
        self.static = static


jax.tree_util.register_pytree_node(
    CustomPytree,
    lambda self: (
        (self.dict, self.attr),
        {"static": self.static},
    ),
    lambda aux, children: CustomPytree(*children, aux["static"]),
)

mytree = CustomPytree(dict={"b": 2}, attr=1, static=1)  # try with dict={"a": 1}

jax.tree_util.tree_leaves(mytree)  # [42, 2, 1]
jax.tree_util.tree_structure(
    mytree
)  # PyTreeDef(CustomNode(CustomPytree[{'static': 1}], [{'a': *, 'b': *}, *]))

# flattennin / unflattening works (example taken from JAX's doc)
value_flat, value_tree = jax.tree_util.tree_flatten(mytree)
value_flat[2] = 3
transformed_structured = jax.tree_util.tree_unflatten(value_tree, value_flat)
print(
    [transformed_structured.dict, transformed_structured.attr]
)  # [{'a': 42, 'b': 4}, 3]

# FAILS with TypeError: Operation undefined, 42 is not a leaf of the pytree.
eqx.tree_at(lambda t: t.attr, mytree, 3)
```

For context : I'm designing an `eqx.Module` with a custom registered pytree as an attribute that have a somewhat similar behavior of setting some dynamic attribute dictionary keys, as the example class above. I've been having trouble using `eqx.tree_at` on the whole module because this custom class is traversed. I could go with a `eqx.tree_at(..., is_leaf=lambda t: isinstance(t, CustomPytree))` to avoid traversing CustomClass but I seem to be missing something important here.

Thanks !

**Comment by lockwo** [comment]
This is failing because `pytree = jtu.tree_map(_LeafWrapper, pytree, is_leaf=is_leaf)` in eqx.tree_at can't make {"a": 42} a LeafWrapper. Specifically, tree map in general, doesn't modify {"a"} (e.g. if you just treemap to cast everything to an array, a will always be an int because when you create the pytree by unflattening you make it so). To be honest, I'm not sure if this is actually a valid pytree, but if it is, it seems hard to build around this. You can't apply any tree operations to it, so maybe you could look at the before and after of a tree map to determine if something is different. 

Conceptually, it makes more sense to have "a" be static if it won't change, or have a be an optional flag in the constructor which defaults to 42, but that might just be me

**Comment by HGangloff** [comment]
As for the validity of the pytree, this is an interesting question. A `tree_map` over such custom pytree won't fail, e.g.

`jax.tree_util.tree_map(lambda x:3 if isinstance(x, int) else x, mytree)`

**Comment by lockwo** [comment]
That's true, tree operations won't necessarily fail directly, but won't act as one might expect. E.g. `jax.tree_map(lambda x: jnp.array(x), mytree)` does not result in a pytree of arrays (42 is still an int)

**Comment by HGangloff** [comment]
When you mention an optional flag, you think about something like:

```
class CustomPytree:
    def __init__(self, dict, attr, static, a_exists=False):
        self.dict = dict
        # preprocess attribute -> always set key "a" to 42
        if not a_exists:
            self.dict["a"] = 42
        self.attr = attr
        self.static = static


jax.tree_util.register_pytree_node(
    CustomPytree,
    lambda self: (
        (self.dict, self.attr),
        {"static": self.static},
    ),
    lambda aux, children: CustomPytree(*children, aux["static"], a_exists=True),
)

mytree = CustomPytree(dict={"b": 2}, attr=1, static=1) 
```

upon which `eqx.tree_at(lambda t: t.attr, mytree, 3)` and  `jax.tree_map(lambda x: jnp.array(x), mytree)` now seem to work.
But this seems hacky for you ?

**Comment by nicolasJouvin** [comment]
Thank you for your answer !

> This is failing because `pytree = jtu.tree_map(_LeafWrapper, pytree, is_leaf=is_leaf)` in eqx.tree_at can't make {"a": 42} a LeafWrapper. Specifically, tree map in general doesn't modify {"a"} (e.g. if you just treemap to cast everything to an array, a will always be an int because when you create the pytree by unflattening you make it so). To be honest, I'm not sure if this is actually a valid pytree because it seems to break some of the assumptions of jax tree utils, but if it is, it seems hard to build around this. You can't apply any tree operations to it, so maybe you could look at the before and after of a tree map to determine if something is different.

 I took a better look at `eqx.tree_at` code and if I understand correctly dict["a"] is supposed to be _LeafWrapper during the reconstruction "equinox unflattening" but my line `self.dict["a"] = 42` completely erases the _LeafWrapper object right ? Interestingly, this error seems reminiscent of [JAX's doc warning about custom pytree initialization.](https://jax.readthedocs.io/en/latest/pytrees.html#custom-pytrees-and-initialization), except Equinox adds the custom class `_LeafWrapper` on top of it.  Like casting  self.dict["a"] = jax.numpy.array(self.dict["a"]) actually fails as because `jnp.array` cannot handle `_LeafWrapper`. 

In the end, one just needs to really be thoughtful in any kind of processing of attributes in `__init__` of custom pytree nodes (including `eqx.Module`) ? 

> Conceptually, it makes more sense to have "a" be static if it won't change, or have a be an optional flag in the constructor which defaults to 42, but that might just be me

I agree with you but in the more complex code underlying this MWE I'd like to keep `dict["a"]` dynamic. Actually, in the real code I had this optional flag behavior, only optionally setting `"a"` to `42` depending on a static attribute. Something like
```
if self.static==1:
  self.dict["a"] = 42
```
I did not include`eqx.tree_at` fails in the same way if the first `mytree` is built with `static=1`. This makes perfect sense and was an error on my part, since this condition in always met in all subsequent calls to `__init__` during unflattening (being pure JAX or equinox). I shall rewrite the test into something like
```
if static == 1 and "a" not in self.dict.keys():
  self.dict["a"] = 42
```

Thank you again, I'm closing this issue for now as I think I have the anwsers I need.
